core/capabilities/syncer_test.go

```diff
@@ -32,9 +38,565 @@ func TestSyncer_CleanStartClose(t *testing.T) {
+					Signer:              randomWord(),
+					P2pId:               capabilityDonNodes[1],
+					HashedCapabilityIds: [][32]byte{triggerCapID, targetCapID},
+				},
+				capabilityDonNodes[2]: {
+					NodeOperatorId:      1,
+					Signer:              randomWord(),
+					P2pId:               capabilityDonNodes[2],
+					HashedCapabilityIds: [][32]byte{triggerCapID, targetCapID},
+				},
+				capabilityDonNodes[3]: {
+					NodeOperatorId:      1,
+					Signer:              randomWord(),
+					P2pId:               capabilityDonNodes[3],
+					HashedCapabilityIds: [][32]byte{triggerCapID, targetCapID},
+				},
+				workflowDonNodes[0]: {
+					NodeOperatorId: 1,
+					Signer:         randomWord(),
+					P2pId:          workflowDonNodes[0],
+				},
+				workflowDonNodes[1]: {
+					NodeOperatorId: 1,
+					Signer:         randomWord(),
+					P2pId:          workflowDonNodes[1],
+				},
+				workflowDonNodes[2]: {
+					NodeOperatorId: 1,
+					Signer:         randomWord(),
+					P2pId:          workflowDonNodes[2],
+				},
+				workflowDonNodes[3]: {
+					NodeOperatorId: 1,
+					Signer:         randomWord(),
+					P2pId:          workflowDonNodes[3],
+				},
+			},
+		},
+	}
+	syncer := newRegistrySyncer(make(services.StopChan), wrapper, registry, dispatcher, lggr, HardcodedDonNetworkSetup{}, mr)
+	require.NoError(t, err)
+
+	dispatcher.On("SetReceiver", fullTriggerCapID, fmt.Sprint(capDonID), mock.AnythingOfType("*remote.triggerSubscriber")).Return(nil)
+	dispatcher.On("SetReceiver", fullTargetID, fmt.Sprint(capDonID), mock.AnythingOfType("*target.client")).Return(nil)
+
+	err = syncer.sync(ctx)
+	require.NoError(t, err)
+
+	_, err = registry.Get(ctx, fullTriggerCapID)
+	require.NoError(t, err)
+
+	_, err = registry.Get(ctx, fullTargetID)
+	require.NoError(t, err)
+}
+
+func TestSyncer_WiresUpClientsForPublicWorkflowDONButIgnoresPrivateCapabilities(t *testing.T) {
+	ctx := tests.Context(t)
+	lggr := logger.TestLogger(t)
+	registry := NewRegistry(lggr)
+	dispatcher := remoteMocks.NewDispatcher(t)
+
+	var pid ragetypes.PeerID
+	err := pid.UnmarshalText([]byte("12D3KooWBCF1XT5Wi8FzfgNCqRL76Swv8TRU3TiD4QiJm8NMNX7N"))
+	require.NoError(t, err)
+	peer := mocks.NewPeer(t)
+	peer.On("UpdateConnections", mock.Anything).Return(nil)
+	peer.On("ID").Return(pid)
+	wrapper := mocks.NewPeerWrapper(t)
+	wrapper.On("GetPeer").Return(peer)
+
+	workflowDonNodes := [][32]byte{
+		pid,
+		randomWord(),
+		randomWord(),
+		randomWord(),
+	}
+
+	capabilityDonNodes := [][32]byte{
+		randomWord(),
+		randomWord(),
+		randomWord(),
+		randomWord(),
+	}
+
+	fullTriggerCapID := "streams-trigger@1.0.0"
+	triggerCapID := randomWord()
+	targetCapID := randomWord()
+	donID := uint32(1)
+	capDonID := uint32(2)
+	// The below state describes a Workflow DON (AcceptsWorkflows = true),
+	// which exposes the streams-trigger and write_chain capabilities.
+	// We expect receivers to be wired up and both capabilities to be added to the registry.
+	mr := &mockReader{
```

Comment:
The setup could do with a helper 😅

Reply:
I agree that this looks daunting, but I'm not sure how this could be abstracted cleanly; exactly what state you produce changes from test to test, so I feel it's better to have control directly over the state struct and build it manually